### ISSUE TITLE
[DFA-arch T2] Latent/low-dim diffusion denoiser + DFA — novel, feasibility-backed (latent ONLY, not pixels)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#156]
**Tier 2 — novel.** Feasibility de-risked by DiffusionBlocks (Shibata et al. 2025, arXiv:2506.14202): block-local, gradient-isolated diffusion training matches end-to-end BP (4x train-memory cut, ~3x faster). But DFA-SPECIFICALLY on the denoiser is unclaimed. Deep activation-heavy net; noise-level decomposition is unusually separable.

## HARD CONSTRAINT (our output-dim lens)
The eps-prediction target has the INPUT's dimensionality. That is small for **latent-space / tabular / time-series / molecular** diffusion but ~2e5 for pixel space -> pixel diffusion defeats DFA exactly as vocab 50257 defeats our LM. **Scope strictly to latent/low-dim.**

## Ask
A minimal **diffusion denoiser** (small MLP/UNet-lite) trained by eps-prediction / score matching on a LOW-DIM target (latent codes, tabular, or a small time-series), with per-layer `--policy chain|dfa`. Compare DFA-denoiser vs BP-denoiser on the generative metric + a frozen-denoiser control.

## Success target
all-DFA denoiser matches BP on the small-output generative task. Backprop-free diffusion is partly claimed (DiffusionBlocks); DFA in the latent/low-dim regime is the open, plantable flag.

## Consuming Tao experiment: F20 (latent-diffusion).
Refs: DiffusionBlocks 2025 (arXiv:2506.14202).

**Comment by OriPekelman** [comment]
Shipped in `2f71c2f` — **the success target is MET and exceeded: DFA beats BP at latent 4, on every seed. And the output-dim lens transfers to a generative objective but bites much earlier than it does on classification.**

`toy train diff` (runner `libexec/toy-train-diff`, engine `lib/toy/llm/engine/diff_engine.rb`, task `lib/toy/io/toy_diff_task.rb`, gate `prep/diff_gate.rb`): a time-conditioned eps-prediction MLP denoiser over a low-dim latent, per-layer `--policy chain|dfa|frozen`. CPU-only. The DFA rule is toy#152's canonical direct-gradient one (with the ⊙f' factor), so a result here is comparable to the MLP anchor straight across — only the output dimension and the loss differ.

Scoped strictly to a low-dim latent, per this ticket's hard constraint.

## 1. The metric is generative, not denoising MSE

You asked to compare "on the generative metric", and that turned out to matter: the two genuinely disagree. The headline is the **energy distance** between `eval_n` ancestrally-sampled points and `eval_n` held-out reals — a proper metric (zero iff the distributions match), so **lower is better** and the success bar inverts. A slightly negative value means "indistinguishable at this sample size" (the estimator is unbiased), not a bug.

## 2. The result

All three arms at **one shared learning rate** (3e-4), 10000 steps, so they differ ONLY in `--policy`:

| latent | seed | BP | **DFA** | frozen |
|---|---|---|---|---|
| 4 | 0 | 0.102 | **0.010** | 0.590 |
| 4 | 1 | 0.184 | **0.135** | 1.084 |
| 4 | 2 | 0.131 | **0.072** | 1.005 |
| 16 | 0 | 0.130 | 0.979 | 9.020 |

**DFA beats BP at latent 4 on every seed** while both crush the frozen control. At latent 16 DFA is 7.5× worse than BP (15–27× at each arm's own best LR across seeds), while still beating frozen 9×.

## 3. The output-dim lens, on a generative objective

Each arm at its own best LR, 10000 steps, seed 0:

| latent | 2 | 4 | 8 | 16 | 32 |
|---|---|---|---|---|---|
| BP | −0.013 | 0.085 | 0.079 | 0.036 | 0.399 |
| DFA | 0.007 | 0.010 | 0.195 | 0.979 | 2.746 |
| frozen | 0.029 | 0.226 | 1.053 | 6.761 | 45.451 |

DFA's energy rises monotonically with the output dim and beats the frozen control at every one. **The boundary sits between 4 and 16 — much lower than the classification lens, where toy#152 was still comfortable at 10 classes.** The lens transfers, but it bites earlier on a generative objective. Worth building into F20's cell design rather than assuming the toy#152 boundary carries over.

At latent 2 the frozen control nearly matches BP (.029 vs −.013), so the lens boundary and the control-can-lose boundary are adjacent — which is why the bar is stated at latent 4. Both ends are gated: the latent-4 win alone reads as "DFA works on diffusion", the latent-16 loss alone as "it does not".

## 4. A schedule bug that was completely silent in the loss

Ho et al.'s betas (1e-4 … 0.02) assume T = 1000. At this lane's T = 100 they leave **abar_T = 0.60** — the forward process never reaches noise, so the ancestral sampler, which starts from pure N(0, I), starts **out of distribution**. At those defaults BP had the *best* denoising MSE (.684) and the *worst* energy distance (29.1, against 4.95 for an untrained net), with a perfectly healthy training curve throughout.

Defaults are now 1e-3 … 0.15 (abar_T = 3.5e-4), and the runner **refuses to start** if abar_T > 0.01, naming the fix. Gated. Flagging it because any diffusion lane built on a short schedule will hit it, and it produces a confident wrong result rather than an error.

## 5. Two methodological facts the gate pins

Getting either wrong **inverts** the finding:
- **Compare arms at lr 3e-4.** BP's own best is 3e-3 (energy .036 at latent 16) but DFA diverges there (4.6e6 at 1e-2), so a 3e-3 cell pits a converged BP against a broken DFA. 3e-4 is the highest rate both arms tolerate — and DFA tolerating less LR than BP is toy#152's finding restated.
- **Use 10000 steps.** At 3000, BP has not converged at 3e-4 and latent 16 reads as parity (BP .802 vs DFA .947).

## 6. Running it

```
toy train diff --steps 10000 --seed 0 --lr 0.0003 --latent 4 --policy chain,chain,chain
toy train diff ... --policy dfa,dfa,dfa
toy train diff ... --policy frozen,frozen,frozen
#  ... and again at --latent 8 / 16 / 32 for the lens
```

`--task single` (one Gaussian instead of the mixture) is this lane's `blobs`: chain −.0020 / dfa −.0011 / frozen −.0012 — a frozen denoiser reproduces it as well as a trained one. Measured, asserted.

Unlike the ssm lane, this one **does** carry `--align-events`: the direct rule makes the DFA update a separate tensor from the chain shadow, so cos(g_dfa, g_bp) is meaningful — and it climbs. That is "align, then memorise" reproduced on a **regression** objective, which as far as I know we had not seen before.

**Comment by OriPekelman** [comment]
Closing: shipped in `2f71c2f` — feat(dfa-arch): toy#156/F20 latent diffusion — DFA BEATS BP at latent 4

Verified by `git log`; the implementing commit names this issue in its subject. Part of the 2026-08-14 issue-state sweep (48 closed, all against a commit in `main`). Reopen if the scope here outran what landed.
